### PR TITLE
feat: logger inheritance

### PR DESCRIPTION
**Pull request by heitorlessa**
**Issue #, if available:** #97 

## Description of changes:

This addresses a common issue when trying to use `Logger` across multiple modules while wanting to:

* **Create a new child Logger** to inherit parent Logger's configuration (handler, formatter, custom keys, sampling)
* **Reuse the same Logger**, where Logger is enriched iteratively across the code base

For this, we've introduced a new parameter, `child`, when initializing a new Logger, and some internal changes while keeping the same UX and backwards compatibility.

* **[Internal]** Direct access to methods, handlers, etc work as before, except that we proxy properties access to `._logger`

## UX

**Key aspects**:

* We support logging inheritance via the `child` parameter, and the `Logger` will be named after `<service>.<caller_file_name>`
* Calling `Logger` with the same `service` name (explicit or via environment variable) returns the same `Logger`
* Changes to either Child `Logger` or Parent `Logger` will be propagated to all instances

### Creating a child Logger

Child logger can be created in any other as long as a parent is created too - This supports customers using Layers or different file names to create child loggers.

```python
# lambda_handler.py
from aws_lambda_powertools import Logger
import collect

logger = Logger(service="payment") # parent Logger just like before, named 'payment'
...

# collect.py
from aws_lambda_powertools import Logger

logger = Logger(service="payment", child=True) # child Logger named 'payment.collect'

# a_new_key will now be available both in this child logger as well as the parent
logger.structure_logs(append=True, a_new_key="value") 
...
```

### Reusing the same Logger

```python
# lambda_handler.py
from aws_lambda_powertools import Logger
import collect

logger = Logger(service="payment") # parent Logger just like before, named 'payment'

# collect.py
from aws_lambda_powertools import Logger

logger = Logger(service="payment") # returns the existing 'payment' Logger
...
```

**Checklist**

 

- [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [x] Update tests
- [ ] Update docs
- [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

**Changes**

- [x] Register logger via `getLogger` to allow inheritance
- [x] Encapsulate logger initialization/config in a new method
- [x] Remove Logger sub-classing
  - [x] Proxy calls to underneath Logger instead
- [x] ~~Create `name` parameter for hierarchy~~
  - [x] ~~Use `service` if `name` is empty~~
- [x] Test on Lambda locally
  - [x] Single Logger, no code changes
  - [x] Two Loggers in different files
  - [x] Additional Logger in a package that is imported on main entry-point
  - [x] Lambda Layer
- [x] Only add a handler if there isn't one
- [x] Only add a handler if parent is root
- [x] Update Formatter to append new keys
- [x] Make `handler` private
- [x] Remove `log_keys`
- [x] Remove duplicate Parameters in docstring
- [x] Double check Debug w/ getattr perf impact
- [x] Docs: Document inheriting Loggers
- [x] Docs: Changelog (doc random logger for tests)
- [x] Test: Logger hierarchy via logger.parent
- [x] Remove `name` parameter
- [x] Allow logger child propagation
- [x] Introduce `child` boolean param
- [x] Don't add a handler nor evaluate sampling if it's a `child` Logger
- [x] Name `child` Loggers based on `<service>.<caller_file_name>` to mimick `getLogger(__name__)`
- [x] Update tests to remove Handler on every test
- [x] Update tests to remove `getRandomLogger` utility function

**Optional**

- [x] Ensure log level input is case insensitive

## Benchmark

> **TL;DR**: For the average use case we have a slight performance decrease, but all negligible in all cases (0.10 nano seconds). 

Quick checks as we've removed `logging.Logger` sub-classing, respect Logging inheritance, and are proxying calls to inner Logger.

### Basic logger

* **Script**: `python -m timeit -s 'from aws_lambda_powertools import Logger; logger = Logger(); logger.info("hello world")'`
* **Before**: 50000000 loops, best of 5: **6.12 nsec** per loop
* **After with Logger proxy object, no sub-classing**: 50000000 loops, best of 5: **6.22 nsec** per loop

### Inheritance

* **Script**: `python -m timeit -s 'from aws_lambda_powertools import Logger; logger = Logger(); l = Logger(); l.info("hello world")'`
* **Same logger**: 50000000 loops, best of 5: **6.28 nsec** per loop
* **Random child logger name + fixup**: 50000000 loops, best of 5: **6.14 nsec** per loop
    - **Script**: `python -m timeit -s 'from aws_lambda_powertools import Logger; logger = Logger(child=True); l = Logger(); l.info("hello world")'`

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws_lambda_powertools/logging/logger.py [hunk 1]
```diff
@@ -112,19 +110,48 @@ def __init__(
         self,
         service: str = None,
         level: Union[str, int] = None,
+        name: str = None,
         sampling_rate: float = None,
         stream: sys.stdout = None,
         **kwargs,
     ):
         self.service = service or os.getenv("POWERTOOLS_SERVICE_NAME") or "service_undefined"
+        self.name = name or self.service
```
Comment (nmoutschen) [inline aws_lambda_powertools/logging/logger.py:119]: I'd have a slight preference to do something like `${name}.${self.service}`, seems easier from an experience point of view to me.

File: aws_lambda_powertools/logging/logger.py [hunk 2]
```diff
@@ -89,18 +90,15 @@ class Logger(logging.Logger):
                 logger.structure_logs(append=True, payment_id=event["payment_id"])
                 logger.info("Hello")
 
-    Parameters
-    ----------
-    logging : logging.Logger
-        Inherits Logger
```
[the remark this answers is not in the document]
Comment (heitorlessa) [reply aws_lambda_powertools/logging/logger.py:95]: No, because I'm proxying all not found attributes directly to the inner logger. It's the same effect of sub-classing for advanced customers who were making changes directly to the logger afterwards , except done right now

File: tests/functional/utility_functions.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import secrets
+
+from aws_lambda_powertools import Logger
+
+
+def get_random_logger(**kwargs):
+    """Return vanilla logger w/ random name"""
+    return Logger(name=secrets.token_urlsafe(), **kwargs)
```
[the remark this answers is not in the document]
Comment (heitorlessa) [reply tests/functional/utility_functions.py:8]: Good call - I accidentally moved all tests to fetch a random Logger to prevent errors/side effects. I'll create one

File: aws_lambda_powertools/logging/logger.py (hunk 1 above)
Comment (heitorlessa) [reply aws_lambda_powertools/logging/logger.py:119]: The `dot` represents inheritance here, and there could be a limit on how many chars you could have in a Logger at some point, hence why either name or service. I actually had it like that but saw this quickly growing when tried something like `name=__name__` in a module

File: aws_lambda_powertools/logging/logger.py [hunk 4]
```diff
@@ -55,6 +54,8 @@ class Logger(logging.Logger):
         service name to be appended in logs, by default "service_undefined"
     level : str, optional
         logging.level, by default "INFO"
+    name: str
+        Logger name, "{service}" by default
```
[the remark this answers is not in the document]
Comment (heitorlessa) [reply aws_lambda_powertools/logging/logger.py:58]: Nope, not required. Just changed the docstring to reflect that - good catch.

File: aws_lambda_powertools/logging/logger.py [hunk 5]
```diff
@@ -112,19 +112,65 @@ def __init__(
...
+
+        self._init_logger(**kwargs)
+
+    def __getattr__(self, name):
+        # Proxy attributes not found to actual logger to support backward compatibility
+        # https://github.com/awslabs/aws-lambda-powertools-python/issues/97
+        return getattr(self._logger, name)
+
+    def _get_logger(self):
+        """ Returns a Logger named {self.service}, or {service.filename} for child loggers"""
```
Comment (alexanderluiscampino) [inline aws_lambda_powertools/logging/logger.py:136]: for consistency: `or {self.service.filename}`

File: aws_lambda_powertools/logging/logger.py [hunk 6]
```diff
@@ -118,7 +118,7 @@ def __init__(
     ):
         self.service = service or os.getenv("POWERTOOLS_SERVICE_NAME") or "service_undefined"
         self.sampling_rate = sampling_rate or os.getenv("POWERTOOLS_LOGGER_SAMPLE_RATE") or 0.0
-        self.log_level = level or os.getenv("LOG_LEVEL") or logging.INFO
+        self.log_level = level or os.getenv("LOG_LEVEL".upper()) or logging.INFO
```
Comment (alexanderluiscampino) [inline aws_lambda_powertools/logging/logger.py:122]: Not sure what is the purpose of  `.upper()` in here since `"LOG_LEVEL"` is explicitly set as all uppercase

Comment (heitorlessa) [reply aws_lambda_powertools/logging/logger.py:122]: `LOG_LEVEL` is the env variable, and one particular customer ended up using "info" as the value for the env var - This solves that. 

In fact, I'm gonna turn this into a private method to cover the explicit `level` too

Comment (alexanderluiscampino) [reply aws_lambda_powertools/logging/logger.py:122]: Gotcha, your explanation makes sense, then you need the .upper outside of the getenv(LOG_LEVEL).UPPER()

**Comment by heitorlessa** [comment]
Just addressed @alexanderluiscampino commentson consistency, and took the opportunity to turn`log level` input to case insensitive to prevent common mistakes for newcomers.

I'll start working on the docs this afternoon or tomorrow, and should release a minor version by EOW

**Comment by heitorlessa** [comment]
Yep it’s fixed in the latest commit ;)
